Books controller tests

mnmallea opened the pull request:
## Summary
- Added tests for books controller index
- Added tests for books controller show

## Trello card
Link [here](https://trello.com/c/OJrz7GSq/22-crear-tests-para-rutas-index-y-show-de-book)

francoprud commented on spec/controllers/books_spec.rb:
```diff
@@ -0,0 +1,53 @@
+      end
+
+      it 'responds with all the books as json' do
+        expect(response.body).to eq serialize_books(books).to_json
```
Here instead you can use:

```ruby
ActiveModel::Serializer::CollectionSerializer.new(
  books, each_serializer: BookSerializer
).to_json
```

francoprud commented on spec/controllers/books_spec.rb:
```diff
@@ -0,0 +1,53 @@
+describe Api::V1::BooksController do
+  include_context 'Authenticated User'
+
+  describe 'GET #show' do
```
Also create a Rent so you can test de `actual_rent` serialization

francoprud commented on spec/controllers/books_spec.rb:
```diff
@@ -0,0 +1,71 @@
+
+      context 'When fetching all the books' do
+        let!(:books) { create_list(:book, 5) }
+        before { get :index }
```
We already talked about this. Small fix

Commit:
Adding tests for BooksController#index

Commit:
Adding books#show tests cases

Commit:
Merge branch 'books-routes' into books-controller-tests
This is because is needed to update the show method

Commit:
Adding filtering tests for /books endpoint

lucasVoboril commented on spec/controllers/shared_context.rb:
```diff
@@ -0,0 +1,7 @@
+shared_context 'Authenticated User' do
```
The name of context should be named starting with `when with or without`
```suggestion
shared_context 'with an authenticated user' do
```

lucasVoboril commented on spec/controllers/shared_context.rb:
```diff
@@ -0,0 +1,7 @@
+shared_context 'Authenticated User' do
+  let(:user) { create(:user) }
```
I prefer to call logged user as current_user
```suggestion
  let(:current_user) { create(:current_user) }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+require 'rails_helper'
+require_relative 'shared_context'
```
I use 
```rb
# spec/rails_helper.rb
Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+require_relative 'shared_context'
+
+describe BooksController do
+  it { is_expected.to route(:get, '/books').to(action: :index) }
```
This is not necessary :wink:

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+describe BooksController do
+  it { is_expected.to route(:get, '/books').to(action: :index) }
+  it { is_expected.to route(:get, '/books/1').to(action: :show, id: 1) }
```
Always avoid using literal numbers

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+  it { is_expected.to route(:get, '/books').to(action: :index) }
+  it { is_expected.to route(:get, '/books/1').to(action: :show, id: 1) }
+
+  describe 'GET #index' do
```
I prefer to read in first line what is the  subject of the current block

```rb
describe 'GET #index' do
  subject(:http_request) { get :index }
```

lucasVoboril replied:
:eyes: you can define the subject at top of describe

Commit:
Fixing books controller tests
This is beacuase there was duplicate test cases

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+  describe 'GET #index' do
+    context 'When an user is authenticated' do
+      include_context 'Authenticated User'
+      let!(:books) { create_list(:book, 5) }
```
If you use build_stubbed instead of create you can improve your performance :wink: [Check it](https://medium.com/@DmytroVasin/speed-up-your-tests-via-build-stubbed-f1926863b3d7)

lucasVoboril replied:
In this case `build_stubbed_list`

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+      include_context 'Authenticated User'
+      let!(:books) { create_list(:book, 5) }
+      context 'When fetching all the books' do
+        before { get :index }
```
```suggestion
        before { http_request }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+      context 'When fetching all the books' do
+        before { get :index }
+
+        it { is_expected.to respond_with :ok }
```
```suggestion
        it { is_expected.to have_http_status :ok }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        end
+      end
+    end
+    context 'When there is not an authenticated user' do
```
```suggestion
    context 'without an authenticated user' do
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+    end
+    context 'When there is not an authenticated user' do
+      context 'When fetching all the books' do
+        before { get :index }
```
This should be your subject :wink: 
```suggestion
        before { http_request }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+    context 'When there is not an authenticated user' do
+      context 'When fetching all the books' do
+        before { get :index }
+        it { is_expected.to respond_with :unauthorized }
```
You can test if your response_body.page to be empty

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        end
+      end
+
+      context 'When params[:author] == Bradford Nolan DDS' do
```
```
context 'when searching by author name' do
let(:author_name) {  'Bradford Nolan DDS' }
        before do
          create_list(:book, 2, author: author_name)
          create_list(:book, 1, author: 'Luis Miguel')
          get :index, params: { author: author_name }
        end
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        end
+      end
+
+      context 'When params[:author] == Bradford Nolan DDS' do
```
Copy paste?

mnmallea replied:
ups :cry:

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+        it { is_expected.to be_paginated }
+        it 'responses with a json with 2 books' do
+          expect(response_body['page'].length).to eq 2
```
:thinking: doesn't exists response_body['count'] ?

mnmallea replied:
yes

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        it 'responses with a json with 2 books' do
+          expect(response_body['page'].length).to eq 2
+        end
+        it 'returns only books with author == Bradford Nolan DDS' do
```
```rb
 it 'returns books filtered by author name' do
authors = response_body['page'].pluck('author')
expect(authors).to(be_all { |author| author.eq(author_name) } )
end
```

mnmallea replied:
Thanks! pluck :heart_eyes: 
Is there any reason to use eq instead of == ?

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        end
+      end
+
+      context 'When params[:title] == La biblia!' do
```
```suggestion
      context 'When filter by title' do
```

lucasVoboril replied:
let(:searched_title) { 'La biblia!' }

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+      context 'When params[:title] == La biblia!' do
+        before do
+          create(:book, title: 'La biblia!')
```
```suggestion
          create(:book, title: searched_title')
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+        before do
+          create(:book, title: 'La biblia!')
+          create(:book, title: 'Harry Potter')
+          get :index, params: { title: 'La biblia!' }
```
```suggestion
          get :index, params: { title: searched_title}
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+  end
+
+  describe 'GET #show' do
+    let(:book) { create(:book) }
```
We are going to create books in contexts :wink: 
```suggestion
    subject(:http_request) { get :show, params: { id: book.id }  }
```

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+  describe 'GET #show' do
+    let(:book) { create(:book) }
+
```
```rb
context 'when searching for an invalid book' do
let(:book) { build_stubbed(:book, id: -1) }
# ...
end
```

mnmallea replied:
No entiendo este.
Lo que se me ocurriría es buscar un libro que no exista y ver si responde un 404

lucasVoboril replied:
Claro! Y un libro con un id negativo nunca te lo va a encontrar, entonces te tiene que devolver un 404

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+    context 'When there is not an authenticated user' do
+      context 'When fetching a book' do
+        before { get :index }
```
Are you testing index or show? :joy:

lucasVoboril commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,161 @@
+
+    context 'When an user is authenticated' do
+      include_context 'Authenticated User'
+      before { get :show, params: { id: book.id } }
```
```suggestion
      before { http_request }
```

Commit:
Fixing books controller tests
- Moving creations to context scope
- Extracting variables
- Renaming 'user' to 'current_user'

Commit:
Adding requires to rails helpers

Commit:
fixing books controller tests

Commit:
Using plug and be all for testing responses

Commit:
Changing subject to htttp_response in books controller tests
